topics/dev/tutorials/interactive-tools/tutorial.md

```diff
@@ -0,0 +1,795 @@
+Let's check this integration on your machine. You can use a VM if you prefer not to modify your machine environment.
+
+> ### {% icon comment %} A note on the OS
+> This part of the tutorial has been tested on Ubuntu and Debian and there is no guaranteed success for other operating systems.
+> If you have another OS on your machine (i.e. Windows or MacOS), you may need to use an Ubuntu virtual machine or perhaps try <a href="https://docs.microsoft.com/en-us/windows/wsl/install" target="_blank">Windows subsystem for Linux</a>).
+{: .comment}
+
+## Docker installation
+
+> ### {% icon hands_on %} Hands-on: Install Docker
+> Install Docker as described on the [docker website](https://docs.docker.com/engine/install/). Click on your distribution name to get specific information.
+{: .hands_on}
+
+## Galaxy installation
+
+> ### {% icon hands_on %} Hands-on: Install Galaxy
+>
+> For Ubuntu:
+> ```sh
+> # Install git to get Galaxy project
+> sudo apt-get install git
+> # Create a working directory and move to it
+> mkdir ~/GxIT && cd ~/GxIT
+> # Get the galaxy project. A new directory named "galaxy" will be created.
+> # This directory contains the whole project
+> git clone https://github.com/galaxyproject/galaxy
+> # Checkout the last stable version (v21.09 as the time of writing)
+> cd galaxy && git checkout v21.09
+> ```
+{: .hands_on}
+
+
+## Galaxy configuration
+
+> ### {% icon hands_on %} Hands-on
+>
+> ```sh
+> cd ~/GxIT/galaxy/config
+> # Create custom config files
+> cat galaxy.yml.interactivetools > galaxy.yml
+> cat job_conf.xml.interactivetools > job_conf.xml
+> cat tool_conf.xml.sample > tool_conf.xml
+> ```
+> In `galaxy.yml`, add the `galaxy_infrastructure_url` parameter to the galaxy section:
+> ```yaml
+> galaxy:
+>   galaxy_infrastructure_url: http://localhost:8080
+> ```
+> This will make galaxy to provide your GxIT using links like <http://your_gxit_identifier.localhost:8080>.
+>
+> Configure the tool panel by adding a section in `~/GxIT/galaxy/config/tool_conf.xml`:
+> ```xml
+>   <section id="interactivetools" name="Interactive tools">
+>     <tool file="interactive/interactivetool_tabulator.xml" />
+>   </section>
+> ```
+> With these lines, Galaxy will create a new section named "Interactive tools" in the tool panel
+> with our interactive tabulator inside.
+> Choose whatever name and id you want as long as the id is unique.
+> And of course, you have no obligation to put your GxITs in this section.
+> You can put them in any section.
+>
+> Finally, copy your GxIT wrapper to the Interactive Tool directory:
+> ```sh
+> cp ~/my_filepath/interactivetool_tabulator.xml ~/GxIT/galaxy/server/tools/interactive/
+> ```
+{: .hands_on}
+
+## Run Galaxy
+
+Go to the Galaxy directory and:
+```sh
+./run.sh
+```
+
+Galaxy is available at [http://localhost:8080/](http://localhost:8080/) and you should be able to use your GxIT.
```

Comment:
Does that really work for ITs that talk back to Galaxy, like Jupyter  ? You'd probably have to either use the host network or `add-host host.docker.internal:host-gateway` in the `docker_run_extra_arguments`

Reply:
@hexylena would this work for Rstudio or would you have to run this differently?

Reply:
it worked for as long as I'd been working on IEs/ITs, did that change? Is it now more isolated? They use the docker gateway address usually to talk back to galaxy, and that worked at the time?

Reply:
I don't know if it helps, but when I have a tool that needs to communicate with galaxy, I use these env vars in the tool xml:
```xml
        <environment_variable name="HISTORY_ID" strip="True">$__history_id__</environment_variable>
        <environment_variable name="GALAXY_URL">${__app__.config.galaxy_infrastructure_url}
  ```
  And in the R side, I do the following:
  ```R
        
    gx <- .self$context$configuration$raw_config$galaxy
    if (is.na(API_KEY <- Sys.getenv("API_KEY", unset=NA))) {
        API_KEY <- gx$api_key
    }
    if (is.na(GALAXY_URL <- Sys.getenv("GALAXY_URL", unset=NA))) {
        GALAXY_URL <- gx$galaxy_url
    }
    if (is.na(HISTORY_ID <- Sys.getenv("HISTORY_ID", unset=NA))) {
        HISTORY_ID <- gx$history_id
    }
    Sys.setenv(GX_API_KEY=API_KEY)
    Sys.setenv(GX_GALAXY_URL=GALAXY_URL)
    GalaxyConnector::gx_init(
        GALAXY_URL=GALAXY_URL,
        API_KEY=API_KEY,
        HISTORY_ID=HISTORY_ID
    )
    
    ...
    
    GalaxyConnector::gx_put(output_file_path)
```
This will work **if your galaxy is exposed to the internet** or your local **DNS can resolve its name**.
If it is not exposed to the internet - for example in a dev environment - you'll not be able to access galaxy with env variables.

Instead, you'll have to write something like :
```
        <environment_variable name="GALAXY_URL">http://172.17.0.1:8080/</environment_variable>
```

An other option, like you said @mvdbeek , is to add some docker extra args, but I don't know how to do it. I remember it was used in GIE, but in GIT I've never tried.


In both cases, galaxy needs to listen on 0.0.0.0 so it is accessible from other networks.
```
uwsgi:
  http: 0.0.0.0:8080
```

In the case of a dev environment, I said to access the galaxy instance with http://localhost:8080 because when we will run the IT, the URL provided by galaxy will be :
subdomain.interactivetool.the_domain_name_you_are_using:port
but the problem is: if you access your galaxy instance with 127.0.0.1:8080, then you'll end up with the url:
 http://subdomain.interactivetool.127.0.0.1:8080/
 which is not a valid url.
 Thus, I always use `localhost:8080`, which is transformed in 
 http://subdomain.interactivetool.localhost:8080/
 
 That's why I suggested to use localhost.